juju/juju
# Internal package

The following moves some of our root packages into a root internal
package. The packages that I'm proposing to move in this PR shouldn't
be used externally. I'd like to move more (see the list below on what we
should keep outside), which can either be done here or in subsequent 
changes.

Cleaning and auditing our codebase to ensure ownership of our packages
will go a long way to state what we support vs. what we want to keep internal
to our codebase. I propose that we don't shove every package into the internal
package as that seems dangerous, so we should strive to ensure a good
balance is achieved.

More of our codebase is being consumed via clients and we should ensure that
the API we offer is public. We should ensure that logical parts of that API are 
understood and that breaking changes are kept to a minimum. This is currently
hard to do, as everything is fair game, so the PR forces us as juju developers
to think about what is on offer.

As an aside, we might be forced to ensure that no internal packages get leaked
into the API client directly. We could easily add a github action to back this up.

----

Obviously, the following go packages should be outside of the internal package:

 - api
 - apiserver
 - cmd
 - controller
 - core
 - environs
 - jujuclient
 - provider
 - rpc
 - version
 - worker

If I've missed any, then do let me know. There are some which are required by 
downstream applications, so we would want to be overly cautious here.

----

Note: If we outwardly want to reject this, then I honestly don't mind either.

## QA steps

This is _just_ a code reorganization, but we should ensure we don't break
anything regarding some sort of hardcoded reflection.

```sh
$ juju bootstrap lxd test --build-agent
```
